service/s3/endpoints.go

```diff
@@ -49,7 +49,9 @@ func resolveDefaultEndpointConfiguration(o *Options) {
 	if o.EndpointResolver != nil {
 		return
 	}
-	o.EndpointResolver = NewDefaultEndpointResolver()
+	o.EndpointResolver = &compatibleEndpointResolver{
```

Comment:
Question: Is this necessary? If this is setting the default config couldn't we just set `EndpointResolverV2` directly and set `EndpointResolver` to `nil`? Maybe I'm missing something 🤔 .

Reply:
good point, i think the reason to actually have it set (rather than nil) is if there is customer code that is relying on that value to actually be set and not `nil`. 
i can see a benefit in removing it, since i dont think its going to be directly used by our code. but this might be outweighed by the reason above. do you see another benefit to setting it to nil?

Reply:
No it just struck me as odd to not be setting resolver v2 in "set defaults" function since that should now be our default. I can see the value in ensuring there is something set though if it's relied upon elsewhere that we aren't aware of.

Reply:
ah right, so i believe the question is why are we setting `Options.EndpointResolverV2` in `finalizeEndpointResolverV2` rather than `resolveDefaultEndpointConfiguration`. and i believe its because theres a concept of "resolving" a setting vs "finalizing" a setting where the former happens before customer mutation via api options, and the latter happens after customer mutation.

https://github.com/aws/smithy-go/blob/main/codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ConfigFieldResolver.java#L114
https://github.com/aws/aws-sdk-go-v2/blob/main/service/s3/api_client.go#L212

and we need `EndpointResolverV2` to be set after all customer mutations to `Options.EndpointResolver` has been finished, so we can detect whether its a customer provided resolver or the one that we provide